File: skillet/src/main/java/dev/kkorolyov/pancake/skillet/ui/entity/EntityList.java
package dev.kkorolyov.pancake.skillet.ui.entity;

import dev.kkorolyov.pancake.skillet.model.GenericEntity;
import dev.kkorolyov.pancake.skillet.model.GenericEntity.EntityChangeEvent;
import dev.kkorolyov.pancake.skillet.model.Model.ModelChangeEvent;
import dev.kkorolyov.pancake.skillet.model.ModelListener;
import dev.kkorolyov.pancake.skillet.model.Workspace;
import dev.kkorolyov.pancake.skillet.model.Workspace.WorkspaceChangeEvent;
import dev.kkorolyov.pancake.skillet.ui.Panel;

import javafx.scene.Node;
import javafx.scene.control.Button;
import javafx.scene.control.ContextMenu;
import javafx.scene.control.Label;
import javafx.scene.control.ScrollPane;
import javafx.scene.layout.VBox;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;

import static dev.kkorolyov.pancake.skillet.decorator.UIDecorator.decorate;

public class EntityList implements Panel, ModelListener<Workspace> {
	private final Map<GenericEntity, Button> buttons = new HashMap<>();
	private final VBox content = new VBox();
	private final VBox root = new VBox(
			decorate(new Label("Entities"))
					.styleClass("panel-header")
					.get(),
			decorate(new ScrollPane(content))
					.compact()
					.get());

	/**
	 * Constructs a new entity list.
	 * @param workspace workspace maintaining entities
	 */
	public EntityList(Workspace workspace) {
		workspace.register(this);
	}

	private void addNewEntities(Workspace workspace) {
		workspace.getEntities().forEach(entity ->
				buttons.computeIfAbsent(entity, k -> {
					Button button = decorate(new Button(k.getName()))
							.maxSize(Double.MAX_VALUE, null)
							.action(() -> workspace.setActiveEntity(k))
							.contextMenu(() -> decorate(new ContextMenu())
									.item("Delete", () -> workspace.removeEntity(k))
									.get())
							.get();

					entity.register((target, event) -> {
						if (EntityChangeEvent.NAME == event) button.setText(target.getName());
					});
					content.getChildren().add(button);
					return button;
				})
		);
	}
	private void removeOldEntities(Workspace workspace) {
		buttons.entrySet().stream()
				.filter(entry -> !workspace.containsEntity(entry.getKey()))
				.collect(Collectors.toSet())
				.forEach(entry -> {
					buttons.remove(entry.getKey());
					content.getChildren().remove(entry.getValue());
				});
	}

	@Override
	public Node getRoot() {
		return root;
	}

	@Override
	public void changed(Workspace target, ModelChangeEvent event) {
		if (WorkspaceChangeEvent.ADD == event) addNewEntities(target);
		else if (WorkspaceChangeEvent.REMOVE == event) removeOldEntities(target);
	}
}

File: skillet/src/main/kotlin/dev/kkorolyov/pancake/skillet/ui/entity/EntityList.kt
package dev.kkorolyov.pancake.skillet.ui.entity

import dev.kkorolyov.pancake.skillet.contextMenu
import dev.kkorolyov.pancake.skillet.item
import dev.kkorolyov.pancake.skillet.model.GenericEntity
import dev.kkorolyov.pancake.skillet.model.GenericEntity.EntityChangeEvent
import dev.kkorolyov.pancake.skillet.model.Workspace
import dev.kkorolyov.pancake.skillet.model.Workspace.WorkspaceChangeEvent
import dev.kkorolyov.pancake.skillet.ui.Panel
import javafx.scene.control.ContextMenu
import javafx.scene.control.Label
import javafx.scene.control.ListView
import javafx.scene.control.cell.TextFieldListCell
import javafx.scene.layout.VBox
import javafx.util.Callback
import javafx.util.StringConverter

/**
 * Displays entities available in the [Workspace].
 * @param workspace workspace containing displayed entities
 */
class EntityList(workspace: Workspace) : Panel {
	private val content: ListView<GenericEntity> = ListView<GenericEntity>().apply {
		styleClass += "panel-content"
		isEditable = true
		cellFactory = Callback {
			object : TextFieldListCell<GenericEntity>() {
				override fun updateItem(item: GenericEntity?, empty: Boolean) {
					super.updateItem(item, empty)
					item?.let { entity ->
						text = entity.name

						contextMenu {
							ContextMenu().apply {
								item("Remove") { workspace.removeEntity(entity) }
							}
						}
					}
				}
			}.apply {
				converter = object : StringConverter<GenericEntity>() {
					override fun toString(o: GenericEntity): String = o.name
					override fun fromString(s: String): GenericEntity = item.apply { name = s }
				}
				prefWidthProperty().bind(it.widthProperty().subtract(4))
			}
		}
		selectionModel.selectedItemProperty().addListener { _, _, newValue ->
			workspace.activeEntity = newValue
		}
	}
	override val root: VBox = VBox().apply {
		styleClass += "panel"
		children += listOf(
				Label("Entities").apply {
					styleClass += "panel-header"
				},
				content)
	}

	init {
		workspace.register({ target, event ->
			when (event) {
				WorkspaceChangeEvent.ADD -> { addNewEntities(target) }
				WorkspaceChangeEvent.REMOVE -> { removeOldEntities(target) }
			}
		})
		addNewEntities(workspace)
	}

	private fun addNewEntities(workspace: Workspace) {
		workspace.entities.filter { entity ->
			content.items.none { it === entity }
		}.forEach {
			it.register { _, event ->
				when (event) {
					EntityChangeEvent.NAME -> content.refresh()
				}
			}
			content.items.add(it)
		}
	}
	private fun removeOldEntities(workspace: Workspace) {
		content.items.filter { it !in workspace }
				.forEach { content.items.remove(it) }
	}
}
